### PR TITLE
Woocommerce instrumentation

### PR DESCRIPTION
**Pull request by abdulwahidsharief**

### CODE REVIEW

File: includes/plugin-instrumentation.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+use Razorpay\Api\Api;
+use Razorpay\Api\Errors;
+
+class TrackPluginInstrumentation
+{
+    public $api;
+
+    public function __construct($key_id, $key_secret)
+    {
+        $this->api = new Api($key_id, $key_secret);
```
Comment (enishant) [inline includes/plugin-instrumentation.php:12]: @abdulwahidsharief Please check if you can use instance `getRazorpayApiInstance` instead re-defining api here.

File: includes/plugin-instrumentation.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            {
+                return ['status' => 'error', 'message' => 'event given as input is not valid'];
+            }
+
+            if (empty($properties) === true or
+                is_array($properties) === false)
+            {
+                return ['status' => 'error', 'message' => 'properties given as input is not valid'];
+            }
+
```
Comment (enishant) [inline includes/plugin-instrumentation.php:30]: @abdulwahidsharief Please include platform = 'WordPress', platform_version, woocommerce_version etc. Initiate these values within constructor and include to existing array $properties.

File: woo-razorpay.php [hunk 3]
```diff
@@ -411,13 +419,123 @@ public function autoEnableWebhook()
...
+            }
+
+            $authEvent = '';
+            $pluginStatusEvent = '';
+
+            $authProperties = [
+                'plugin_name'             => get_plugin_data(__FILE__)['Name'],
+                'plugin_version'          => get_plugin_data(__FILE__)['Version'],
+                'event_timestamp'         => time(),
+                'woocommerce_version'     => WOOCOMMERCE_VERSION,
```
Comment (enishant) [inline woo-razorpay.php:478]: @abdulwahidsharief Please include all woocommerce_version within [rzpTrackSegment](https://github.com/razorpay/razorpay-woocommerce/pull/305/files#r911386538) method at class TrackPluginInstrumentation

File: includes/plugin-instrumentation.php (hunk 1 above)
Comment (abdulwahidsharief) [reply includes/plugin-instrumentation.php:12]: for some events have to get key and secret submitted by user so creating API object here

File: includes/plugin-instrumentation.php (hunk 2 above)
Comment (abdulwahidsharief) [reply includes/plugin-instrumentation.php:30]: @enishant done

File: woo-razorpay.php (hunk 3 above)
Comment (abdulwahidsharief) [reply woo-razorpay.php]: @enishant added as default properties for all events

**Review by enishant — COMMENTED** [review]
LGTM

File: woo-razorpay.php [hunk 4]
```diff
@@ -411,13 +419,117 @@ public function autoEnableWebhook()
 
         }
 
+        public function pluginInstrumentation()
+        {
+            $trackObject = new TrackPluginInstrumentation($_POST['woocommerce_razorpay_key_id'], $_POST['woocommerce_razorpay_key_secret']);
+
+            if (empty($_POST['woocommerce_razorpay_key_id']) or
```
Comment (ramth05) [inline woo-razorpay.php:426]: @abdulwahidsharief , move check at the start of function before these gets used anywhere.

File: woo-razorpay.php [hunk 5]
```diff
@@ -596,7 +708,21 @@ private function getRedirectUrl($orderId)
          */
         protected function getRazorpayPaymentParams($orderId)
         {
-            
+            $getWebhookFlag =  get_option('webhook_enable_flag');
```
Comment (ramth05) [inline woo-razorpay.php:711]: @abdulwahidsharief , is this auto enable again needs to added in this ?

File: woo-razorpay.php [hunk 6]
```diff
@@ -2073,3 +2199,76 @@ function disable_coupon_field_on_cart($enabled)
...
+// plugin deactivation hook
+function razorpayPluginDeactivated()
+{
+    $paymentSettings = get_option('woocommerce_razorpay_settings');
+
+    $trackObject  = new TrackPluginInstrumentation($paymentSettings['key_id'], $paymentSettings['key_secret']);
+
+    $api = new Api($paymentSettings['key_id'], $paymentSettings['key_secret']);
+
+    $orderCount = $api->request->request('GET', 'orders')['count'];
```
Comment (ramth05) [inline woo-razorpay.php:2231]: @abdulwahidsharief , why we are fetching the order count here ? if the order already created by another plugin or manually?

File: woo-razorpay.php (hunk 5 above)
Comment (abdulwahidsharief) [reply woo-razorpay.php:711]: @ramth05 no after git conflict resolved it is displaying as new line

File: woo-razorpay.php (hunk 6 above)
Comment (abdulwahidsharief) [reply woo-razorpay.php:2231]: @ramth05 have to check if merchant is a transacting user

**Review by ramth05 — COMMENTED** [review]
LGTM, @abdulwahidsharief please test it properly with in team, before merge.

**Review by enishant — APPROVED** [review]
LGTM

**Pull request merged**